blocks/algorithms/__init__.py

```diff
@@ -0,0 +1,239 @@
+
+from blocks.graph import ComputationGraph
+
+
+class TrainingAlgorithm(object):
+    """Base class for training algorithms.
+
+    A training algorithm object has a simple life-cycle.
+    First it is initialized by calling its :meth:`initialize` method.
+    At this stage, for instance, Theano functions can be compiled.
+    After that the :meth:`process_batch` method is repeatedly
+    called with a batch of training data as a parameter.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def initialize(self):
+        """Initialize the training algorithm."""
+        pass
+
+    @abstractmethod
+    def process_batch(self, batch):
+        """Process a batch of training data.
+
+        Attributes
+        ----------
+        batch : dict
+            A dictionary of (source name, data) pairs.
+
+        """
+        pass
+
+
+class DifferentiableCostMinimizer(TrainingAlgorithm):
+    """Minimizes a differentiable cost given as a Theano expression.
+
+    Very often the goal of training is to minimize the expected value of a
+    Theano expression. Batch processing in this cases typically consists of
+    running a (or a few) Theano functions.
+    :class:`DifferentiableCostMinimizer` is the base class for such
+    algorithms.
+
+    Parameters
+    ----------
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned. If ``None``, all shared variables of
+        `cost` computation graph will be considered parameters.
+
+    Attributes
+    ----------
+    updates : list of (shared variable, Theano expression) tuples
+        Updates to be done for every batch. It is required that the
+        updates are done using the old values of optimized parameters.
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Notes
+    -----
+        Changing `updates` attribute or calling `add_updates` after
+        the `initialize` method is called will have no effect.
+
+    .. todo::
+
+        Some shared variables are not parameters (e.g. those created by
+        random streams)
+
+    .. todo::
+
+        Due to a rather premature status of the :class:`ComputationGraph`
+        class the parameter used only inside scans are not fetched
+        currently.
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, cost, params=None):
+        self.cost = cost
+        self.params = (params if params
+                       else ComputationGraph(cost).get_shared_variables())
+        self._cost_computation_graph = ComputationGraph(self.cost)
+        self._updates = []
+
+    @property
+    def inputs(self):
+        """Return inputs of the cost computation graph.
+
+        Returns
+        -------
+            list of Theano variables
+
+        """
+        return self._cost_computation_graph.inputs
+
+    @property
```

Comment:
Right, it does not. It is here mostly because I wanted to have a computation graph object as a private attribute and to have shortcuts to access it.  Maybe I will refactor this place later.